### ISSUE TITLE
Focus management for keyboard users

### ISSUE THREAD

**Issue by AleksueiR** [fgpv-vpgf/fgpv-vpgf#876]
PR #811 (issue #52) added keyboard user detection (by monitoring key events) which sets the global `keyboard` style  (this will show all `settings` toggles in legend).

I suggest that we avoid any focus manipulation unless we detect a keyboard user. Changing focus may seem inconsistent to mouse users and can contradict user expectations.

Here's an example. `Land Pub`'s settings panel is opened; user closes it; focus is reset to the first element on the toc panel (which is the first layer). The user sees focus moving unexpectedly from the second layer to the first. (Ideally, for keyboard users, the focus should be restored to the second layer, but this would require tracking exit/enter focus for each panel).

![focus_rogue](https://cloud.githubusercontent.com/assets/2285779/16778235/1e98739c-483b-11e6-9849-1f4291d3fb1d.gif)
 
Related #815

**Comment by AleksueiR** [comment]
Things discussed in the pod:
- focus management is made somewhat difficult as the focus can escape the viewer's container when a focused element is destroyed (removing a layer would destroy the "remove" button the user is focused on) and browser sets the focus on the body of the page
- the focus also escapes when a drop down menu is opened since the menu control is rendered outside of the viewer's container, at the bottom of the page
- recapturing focus is complicated because it's not clear why the focus was lost (user intentionally leaving the viewer or through destroyed elements)

Suggestions:
- in the "keyboard-only" mode, when the focus enters the viewer's container, it is not allowed to leave without an explicit user action (pressing a key combination or clicking somewhere outside the viewer's container); if the focus is lost, it is automatically restored to the previous known element (in some cases as in layer reordering, to the same element after it is rendered again)
- when tabbing through the page, the viewer's container represents a single focusable element unless "activated" by the user which captures the focus inside the viewer's container
- when the viewer's container is initially focused on, an info popup is shown inside the container informing the user how to start interacting with the viewer (same information is provided through aria attributes for screen readers; information on how to exit the viewer can be similarly provided)
![](https://i.imgur.com/VTJ6fq2.png)

**Comment by milesap** [comment]
This feature is taking slightly longer than expected for a few reasons I didn't account for in the estimate:

- focus management now takes full control of native browser focus.
- taking control of focus requires us to alert the user and instruct them on how to stop this behavior.
- focus manager is much more autonomous than before; there is little need to micro manage its behavior across the application. While it takes longer to automate, it does save time down the road.

This feature also fixes issues #825 and #665.
